l1-contracts/src/core/libraries/validator-selection/ValidatorSelectionLib.sol

```diff
@@ -36,23 +40,30 @@ library ValidatorSelectionLib {
   /**
    * @notice  Performs a setup of an epoch if needed. The setup will
    *          - Sample the validator set for the epoch
-   *          - Set the seed for the epoch
-   *          - Update the last seed
-   *
-   * @dev     Since this is a reference optimising for simplicity, we store the actual validator set in the epoch structure.
-   *          This is very heavy on gas, so start crying because the gas here will melt the poles
-   *          https://i.giphy.com/U1aN4HTfJ2SmgB2BBK.webp
+   *          - Set the seed for the next epoch
    */
-  function setupEpoch(StakingStorage storage _stakingStore) internal {
-    Epoch epochNumber = Timestamp.wrap(block.timestamp).epochFromTimestamp();
+  function setupEpoch(StakingStorage storage _stakingStore, Epoch _epochNumber) internal {
     ValidatorSelectionStorage storage store = getStorage();
-    EpochData storage epoch = store.epochs[epochNumber];
 
-    if (epoch.sampleSeed == 0) {
-      epoch.sampleSeed = getSampleSeed(epochNumber);
-      epoch.nextSeed = store.lastSeed = computeNextSeed(epochNumber);
-      epoch.committee = sampleValidators(_stakingStore, epochNumber, epoch.sampleSeed);
+    // Get the sample seed for this current epoch.
+    uint224 sampleSeed = getSampleSeed(_epochNumber);
+
+    // If no sample seed is set, we are in a genesis state, we set the sample seed to max and push it into the store
+    if (sampleSeed == 0) {
```

Comment:
Should we just set it up at genesis instead? As part of the construction? I might be forgetting something, but would that not at least get rid of this special case in the function that we will continously call? 🤷

Reply:
Yep, lets